### PR TITLE
Fixes 67: Adding test coverage for louis-crawler

### PR DESCRIPTION
**Pull request by melanie-fressard**
issue #67

### CODE REVIEW

**Comment by melanie-fressard** [comment]
Current test coverage is very good !
image

**Comment by melanie-fressard** [comment]
Coverage got up to 95%
image

File: tests/test_db_crawler.py [hunk 1]
```diff
@@ -53,24 +52,77 @@ def test_fetch_crawl_row_by_postgresql_url(self):
...
+            crawler.store_embedding_item(cursor, item)
+            self.connection.rollback()
+        self.assertEqual(item["token_id"], "be612259-9b52-42fd-8d0b-d72120efa3b6")
+
+    def test_fetch_crawl_ids_without_chunk(self):
+        """Test fetching crawl IDs without a chunk."""
+        with db.cursor(self.connection) as cursor:
+            id = crawler.fetch_crawl_ids_without_chunk(cursor)
+            self.connection.rollback()
+        self.assertEqual(id, [])
```
Comment (k-allagbe) [inline tests/test_db_crawler.py:108]: This assertion isn't adequately testing `fetch_crawl_ids_without_chunk`. A better test would be to first populate the table with a mix of crawls, some with chunk IDs and some without. Then, we should verify that the function's results match exactly the crawls that don't have chunk IDs, before performing a rollback.

File: tests/test_db_crawler.py [hunk 2]
```diff
@@ -53,24 +52,77 @@ def test_fetch_crawl_row_by_postgresql_url(self):
...
+
+    def test_fetch_crawl_row(self):
+        """Test fetching a crawl row."""
+        with db.cursor(self.connection) as cursor:
+            row = crawler.fetch_crawl_row(cursor, "https://inspection.canada.ca/a-propos-de-l-acia/structure-organisationnelle/mandat/fra/1299780188624/1319164463699")
```
Comment (k-allagbe) [inline tests/test_db_crawler.py:113]: This data might not exist in the future, in which case this test will fail. A more suitable test would be to first populate the relevant tables with test items, perform the fetch, then assert the result matches the test items.

`fetch_crawl_row` also has multiple paths and edge cases worth testing.

File: tests/test_db_crawler.py [hunk 3]
```diff
@@ -53,24 +52,77 @@ def test_fetch_crawl_row_by_postgresql_url(self):
...
+            self.connection.rollback()
+        self.assertEqual(row['title'], "Mandat - Agence canadienne d'inspection des aliments")
+        
+    def test_fetch_chunk_token_row(self):
+        """Test fetching a chunk token row."""
+        with db.cursor(self.connection) as cursor:
+            row = crawler.fetch_chunk_token_row(cursor, "469812c5-190c-4e56-9f88-c8621592bcb5")
```
Comment (k-allagbe) [inline tests/test_db_crawler.py:120]: This data might not exist in the future, in which case this test will fail. A more suitable test would be to first populate the relevant tables with test items, perform the fetch, then assert the result matches the test items.

**Review by k-allagbe — CHANGES_REQUESTED** [review]
- Impressive 95% test coverage overall. Testing edge cases would bring it up even more.
- Be careful not to use real data in tests. Instead, a good strategy would be to populate the tables with test data before each test and rollback after. There are other strategies you can explore as well.
- Make sure assertions actually assert the effects of the functions.

**Comment by melanie-fressard** [comment]
Pushing it to 96% :
image
Will apply correction requested by @k-allagbe soon :)

File: tests/test_db_crawler.py [hunk 4]
```diff
@@ -53,24 +52,77 @@ def test_fetch_crawl_row_by_postgresql_url(self):
...
+                "url": "https://example.com",
+                "title": "Example",
+                "html_content": "<html><body>This is an example.</body></html>",
+                "lang": "en",
+                "last_crawled": "2022-01-01",
+                "last_updated": "2022-01-01"
+            }
+            crawler.store_crawl_item(cursor, item)
+            self.connection.rollback()
+        self.assertEqual(item["url"], "https://example.com")
```
[the remark this answers is not in the document]
Comment (melanie-fressard) [reply tests/test_db_crawler.py]: done

File: tests/test_db_crawler.py [hunk 5]
```diff
@@ -53,24 +52,77 @@ def test_fetch_crawl_row_by_postgresql_url(self):
...
+            }
+            crawler.store_chunk_item(cursor, item)
+            self.connection.rollback()
+        self.assertEqual(item["title"], "À propos de l'ACIA - Agence canadienne d'inspection des aliments")
```
[the remark this answers is not in the document]
Comment (melanie-fressard) [reply tests/test_db_crawler.py]: done

File: tests/test_db_crawler.py [hunk 6]
```diff
@@ -53,24 +52,77 @@ def test_fetch_crawl_row_by_postgresql_url(self):
...
-
     def test_fetch_chunk_id_without_embedding(self):
         """sample test to check if fetch_chunk_id_without_embedding works"""
         with db.cursor(self.connection) as cursor:
             cursor.execute(test.embedding_table.format(embedding_model='test-model'))
             rows = crawler.fetch_chunk_id_without_embedding(cursor, 'test-model')
             _entity_id = rows[0]
             self.connection.rollback()
+
+    def test_store_chunk_item(self):
```
[the remark this answers is not in the document]
Comment (melanie-fressard) [reply tests/test_db_crawler.py:63]: broke down store_chunk_item into tinier functions
still need to test if there is a case with none

File: tests/test_db_crawler.py [hunk 7]
```diff
@@ -53,24 +52,78 @@ def test_fetch_crawl_row_by_postgresql_url(self):
...
...7317,24789,266,11,9131,11,11376,11,4046,6414,51084,951,97035,13,8245,22139,64829,29696,409,11692,1880,409,85182,77,685,328,5164,409,80080,423,944,59307,80080,11,17889,1354,5860,24985,3946,11,62163,409,5790,325,625,3808,951,35030,3557,1880,3930,586,13,51097,4972,35933,44564,20392,3869,326,58591,73511,7769,3136,13,3744,268,2850,1900,5856,288,9952,3625,447,2053,17317,58673,484,2439,5019,65827,268,404,1880,5201,261,3625,7617,288,409,77463,38450,12267,1880,1097,73511,15171,1208,9313,1880,1208,6225,31617,8082,951,1615,316,18543,1759,13,5856,288,8666,266,22589,1219,13109,8666,50848,294,6,4683,15916,11,1219,13109,8666,1413,3930,49904,265,11,43252,89781,288,1765,3625,13826,25108,4978,409,51304,13,29124,6414,51084,951,97222,1880,951,3600,45629,288,5019,38682,404,15907,22639,9952,3625,2027,38647,11,3625,1615,316,18543,1759,11,326,6,485,592,7379,1880,3625,46106,31957,1821,13,20915,21066,409,1208,26965,13109,44564,3057,1557,409,1208,26965,13109,38682,1001,12267,13,2998,2842,40280,1880,82620,27220,18042,283,35573,514,82620,27220,1880,326,91655,11323,266,2428,13,2998,29033,6672,51097,737,2727,392,645,432,1137,625,3808,1765,3625,737,2727,392,645,1880,46106,21744,10515,5512,409,326,6,1741,5987,13],
```
Comment (JolanThomassin) [inline tests/test_db_crawler.py:97]: is this normal? does all this number have to be in the test_code directly?

**Review by JolanThomassin — CHANGES_REQUESTED** [review]

File: tests/test_db_crawler.py (hunk 7 above)
Comment (melanie-fressard) [reply tests/test_db_crawler.py:97]: well it needs a vector of a certain dimension so I put one of the required lenght but yeah it's not very pretty, I'll try to look into a random generator

Comment (melanie-fressard) [reply tests/test_db_crawler.py]: done

File: tests/test_db_crawler.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (melanie-fressard) [reply tests/test_db_crawler.py:63]: done

**Comment by melanie-fressard** [comment]
To generate documentation based on docstring, python -m pydoc -p 1234 will open a localhost on port 1234 to display the documentation on html pages. Unfortunately, it do not detect the tests module and therefore is not doing any documentation on that despite dosctrings.

**Review by JolanThomassin — APPROVED** [review]

**Comment by melanie-fressard** [comment]
Tried to apply more correction but cannot because I have trouble with adding test data to the database. As it is true that test-data should be added for better testing, I would suggest to start from here if anyone is assigned to this issue. 
Otherwise, pushed tests to 97% :
image

**Comment by rngadam** [comment]
@k-allagbe can you follow up yourself and get this PR in?

**Comment by k-allagbe** [comment]
> @k-allagbe can you follow up yourself and get this PR in?

On it.
